File: frontend/src/utils/textMetrics.js
export const SOFT_LIMIT_CHARACTER_COUNT = 200000

const INVISIBLE_CHARACTER_PATTERN = /[\s\u200B-\u200D\uFEFF]/g

export const countEffectiveCharacters = (text) => (
  String(text || '').replace(INVISIBLE_CHARACTER_PATTERN, '').length
)

export const exceedsSoftLimit = (text, limit = SOFT_LIMIT_CHARACTER_COUNT) => (
  countEffectiveCharacters(text) > Number(limit || SOFT_LIMIT_CHARACTER_COUNT)
)

File: application/services/v1/text_metrics.py
"""V1.1 text metrics baseline."""

from __future__ import annotations

import re


SOFT_LIMIT_CHARACTER_COUNT = 200_000
INVISIBLE_CHARACTER_PATTERN = re.compile(r"[\s\u200b-\u200d\ufeff]+", re.UNICODE)


def count_effective_characters(text: object) -> int:
    """Count visible body characters after removing invisible whitespace."""
    return len(INVISIBLE_CHARACTER_PATTERN.sub("", str(text or "")))


def exceeds_soft_limit(text: object, limit: int = SOFT_LIMIT_CHARACTER_COUNT) -> bool:
    resolved_limit = int(limit or SOFT_LIMIT_CHARACTER_COUNT)
    return count_effective_characters(text) > resolved_limit
